File: php/life/world.php
<?php

error_reporting(E_ALL);

require_once 'cell.php';
require_once 'setup.php';

class World
{

    private $EMPTY_CELL;
    private $param;
    private $step = 0;
    private $grid = array();    // Implementing as hashed array provides more flexibility
    private $neighbours_at = array(
        array(-1, 1),  array(0, 1),  array(1, 1), // above
        array(-1, 0),  /*  self  */  array(1, 0), // beside
        array(-1, -1), array(0, -1), array(1, -1) // below
    );

    function __construct($predefined = false)
    {
        $this->EMPTY_CELL = new Cell;
        $this->param = new Setup;
        if ($predefined) {
            $this->predefined();
        } else {
            $this->prepopulate();
        }
    }

    public function to_string()
    {
        $output = "";
        for ($y = 1; $y <= $this->param->height; $y++) {
            for ($x = 1; $x <= $this->param->width; $x++) {
                $output .= $this->cell_at($x, $y)->to_string();
            }
            $output .= "\n";
        }
        return $output;
    }

    public function calculate()
    {
        for ($y = 1; $y <= $this->param->height; $y++) {
            for ($x = 1; $x <= $this->param->width; $x++) {
                $cell = $this->cell_at($x, $y);
                $neighbours = $this->count_neighbours($x, $y);
                if ($cell->is_alive()) {
                    if (!array_key_exists("s$neighbours",$this->param->survive_values)) {
                        $cell->update_state(false);
                    }

                } else {
                    if (array_key_exists("b$neighbours",$this->param->birth_values)) {
                        $cell->update_state(true);
                    }
                }
            }
        }

        for ($x = 1; $x <= $this->param->width; $x++) {
            for ($y = 1; $y <= $this->param->height; $y++) {
                $this->cell_at($x, $y)->refresh();
            }
        }

        $this->step++;
    }

    public function step()
    {
        return $this->step;
    }

    private function count_neighbours($x, $y)
    {
        $total_neighbours = 0;
        foreach ($this->neighbours_at as $offset) {
            if ($this->cell_at($x + $offset[0], $y + $offset[1])->is_alive()) {
                $total_neighbours++;
            }
        }
        return $total_neighbours;
    }

    private function prepopulate()
    {
        for ($x = 1; $x <= $this->param->width; $x++) {
            for ($y = 1; $y <= $this->param->height; $y++) {
                $alive = (rand(1, 100) <= $this->param->spawn_percent);
                $this->add_cell($x, $y, $alive);
            }
        }
    }

    private function predefined()   // One oscillator to test that we are working
    {
        for ($x = 1; $x <= $this->param->width; $x++) {
            for ($y = 1; $y <= $this->param->height; $y++) {
                $alive = false;
                if ($x==5 && ($y == 4 || $y == 5 || $y == 6)) {
                    $alive = true;
                }
                $this->add_cell($x, $y, $alive);
            }
        }
    }

    private function cell_at($x, $y)
    {
        $gridref = $this->grid_reference($x, $y);
        if (array_key_exists($gridref,$this->grid)) {
            return $this->grid[$this->grid_reference($x, $y)];
        } else {
            return $this->EMPTY_CELL;
        }
    }

    private function add_cell($x, $y, $alive)
    {
        $this->grid[$this->grid_reference($x, $y)] = new Cell($alive);
    }

    private function grid_reference($x, $y)
    {
        $x = $this->wrap_coord($x, 1, $this->param->width, $this->param->wrapx);
        $y = $this->wrap_coord($y, 1, $this->param->height, $this->param->wrapy);
        return "$x|$y";
    }

    private function wrap_coord($val, $min, $max, $wrap_enabled) {
        if ($val < $min) {
            $val = $wrap_enabled ? $max - $val : $min - 1;
        } elseif ($val > $max) {
            $val = $wrap_enabled ? $val - $max : $max + 1;
        }
        return $val;
    }
}

File: python/life/world.py
from life import param
from life import cell
import random

class World:
    neighboursAt = [
        [-1, 1], [0, 1], [1, 1],    # above
        [-1, 0],         [1, 0],    # beside
        [-1, -1], [0, -1], [1, -1]  # below
    ]

    def __init__(self, useTestWorld = False):
        self.EMPTY_CELL = cell.Cell()
        self.step = 0
        self.grid = {}
        self.param = param.Param()
        if useTestWorld:
            self._predefined()
        else:
            self._prepopulate()

    def _predefined(self):
        for x in range(1, self.param.width+1):      # range(1,j) steps from i to j-1
            for y in range(1, self.param.height+1):
                alive = True if x == 5 and (4 <= y <= 6) else False
                self._addCell(x, y, alive)

    def _prepopulate(self):
        for x in range(1, self.param.width+1):
            for y in range(1, self.param.height+1):
                alive = True if random.randint(0, 99) <= self.param.spawnPercent else False
                self._addCell(x, y, alive)
        
        
    def _addCell(self, x, y, alive):
        self.grid[self._gridReference(x, y)] = cell.Cell(alive)

    def _gridReference(self, x, y):
        x = self._wrapCoord(x, 1, self.param.width, self.param.wrapX)
        y = self._wrapCoord(y, 1, self.param.height, self.param.wrapY)
        return str(x) + "|" + str(y)
    
    def _wrapCoord(self, val, min, max, wrapEnabled):
        if val < min:
            val = max - val if wrapEnabled else min - 1
        elif val > max:
            val = val - max if wrapEnabled else max + 1
        return val

    def calculate(self):
        for y in range(1, self.param.height+1):
            for x in range(1, self.param.width+1):
                myCell = self._cellAt(x, y)
                neighbours = self._countNeighbours(x, y)
                if myCell.isAlive():
                    if not f"s{neighbours}" in self.param.surviveValues:
                        myCell.updateState(False)
                else:
                    if f"b{neighbours}" in self.param.birthValues:
                        myCell.updateState(True)

        for y in range(1, self.param.height+1):
            for x in range(1, self.param.width+1):
                self._cellAt(x, y).refresh()

        self.step += 1

    def _cellAt(self, x, y):
        gridRef = self._gridReference(x, y)
        return self.grid[gridRef] if gridRef in self.grid else self.EMPTY_CELL

    def _countNeighbours(self, x, y):
        totalNeighbours = 0
        for offset in self.neighboursAt:
            if self._cellAt(x + offset[0], y + offset[1]).isAlive():
                totalNeighbours += 1
        return totalNeighbours

    def __str__(self):
        output = ""
        for y in range(1, self.param.height+1):
            for x in range(1, self.param.width+1):
                output += str(self._cellAt(x, y))
            output += "\n"
        return output
